### PR TITLE
[tcli-3] Dashboard upgrade notification.

### LINKED ISSUE #15: Notification of updates in the WP dashboard
It would be nice to be able to update the plugin automatically via the WP dashboard when a new version is released, as many other plugins can be (including the Toolset suite).

### PR DESCRIPTION
**Pull request by baizmandesign**
@zaantar Here's a draft of the dashboard notification function (#15).

There's one problem that I couldn't figure out. When the plugin is updated in the dashboard, the version number disappears instead of being updated. See attached screenshots.

screenshot 2021-03-28 at 23 09 32
screenshot 2021-03-28 at 23 09 43

I might also be inclined to change many of the functions' scopes from public to private. And perhaps rename some of the variables, which are still the names from [the Smashing Magazine article](https://www.smashingmagazine.com/2015/08/deploy-wordpress-plugins-with-github-using-transients/) you referenced.

### CODE REVIEW

**Comment by zaantar** [comment]
Thanks a lot for this, @baizmandesign! Sure, if you can reduce the public interface of the Updater class, that will be great. Do you want me to look into the issue with the version number disappearing, or do you plan to investigate further?

**Comment by baizmandesign** [comment]
@zaantar Yes, I can adjust the method scopes. Regarding the version number issue, I did some investigation after submitting this PR. I looked at the source of [this plugin](https://github.com/afragen/github-updater/) and [this plugin](https://github.com/YahnisElsts/plugin-update-checker). They both appear to reference a WP class named [Plugin_Upgrader](https://developer.wordpress.org/reference/classes/plugin_upgrader/) (and [Plugin_Upgrader_Skin](https://developer.wordpress.org/reference/classes/plugin_upgrader_skin/)), but the workings of these core WP classes are, frankly, over my head. I'm not even sure what my next step would be aside from posting a question on StackOverflow. If you could take a look, I'd appreciate it. It also appeared to be the case that the two plugins I just mentioned didn't make use of the `upgrader_post_install` filter (perhaps because they used the Plugin_Upgrader class?).

**Comment by zaantar** [comment]
@baizmandesign sure thing, I will take a look! You've already done a tremendous job here. But as always, please be patient with me, I might be able to find some time next week, hopefully 😅

**Comment by baizmandesign** [comment]
@zaantar No rush on this, Jan! For me, the updating works and the disappearance of the version number is a cosmetic issue, but for others it's a usability issue and ought to be corrected.

I just added another commit. It cleaned up the original code by altering the method scopes, adding comments, renaming variables more descriptively, and adding a bunch of getters and setters.

**Comment by zaantar** [comment]
@baizmandesign FYI, I've done some polishing to your code and light testing, and I couldn't reproduce the issue with the version number not being available. I hope that's actually good news, perhaps it was just some glitch during the development :) 

Anyways, I'll send this to our QA team, so we should know for sure.

**Comment by baizmandesign** [comment]
@zaantar That's odd, but great, to hear that issue didn't surface for you. For me, it happened predictably and every single time I upgraded the plugin.

(The nice thing about this feature, by the way, isn't that one can upgrade it through the dashboard—it's that the plugin itself can be updated through WP CLI!)

**Comment by zaantar** [comment]
@baizmandesign It's also possible that I've fixed it while making changes to your code 😅 We'll see. And yes, you're right, that's pretty cool :) Most importantly, it's consistent with all other plugins now.

**Comment by baizmandesign** [comment]
@zaantar FYI, I tried testing your commit this morning to see if the issue I'd reported regarding the disappearing version number was corrected through your revisions, and unfortunately the version number still disappeared for me. Just wanted to let you know. (There's no mention of this issue in the comments on the Smashing Magazine article, either, so perhaps there's something funky with my WP install.)

File: toolset-cli.php [hunk 1]
```diff
@@ -1,16 +1,19 @@
...
 /*
 Plugin Name: Toolset CLI
-Description: WP-CLI commands for Toolset plugins
+Plugin URI: https://github.com/OnTheGoSystems/toolset-cli
+Description: WP-CLI commands for Toolset plugins.
 Author: OnTheGoSystems
-Author URI: http://www.onthegosystems.com
-Version: 1.0.3
+Author URI: https://onthegosystems.com
+Version: 1.0.2
```
Comment (baizmandesign) [inline toolset-cli.php]: @zaantar This version should be 1.0.3, not 1.0.2.

**Comment by zaantar** [comment]
@baizmandesign, thanks a lot, good catch. I must have accidentally committed this change while testing and fixing some glitches. Should be ok now.

**Comment by baizmandesign** [comment]
@zaantar Thanks for cleaning up the code and merging the PR!

We didn't discuss the workflow for forcing the update nag in the WP dashboard: 
1. on the master branch, update the version number in the header of the plugin file (/toolset-cli.php)
2. commit and push to GitHub 
3. create a new release via the GitHub website*

*I'm unsure if you can create a new release just by creating a new tag on the local master branch and pushing the tag; you're probably more familiar with GitHub's workflows than I am.

Anyhow, this should probably be documented somewhere.

**Comment by zaantar** [comment]
@baizmandesign, you're welcome :)

The release, well, it is basically how you describe: Raise the version number in the plugin header, then commit and create a corresponding tag and push it to GitHub. Creating a release on GitHub doesn't happen automatically just by creating a new tag, but we have this [handled by GitHub Actions](https://github.com/OnTheGoSystems/toolset-cli/blob/master/.github/workflows/release.yml).

As for (public) documentation, I don't think this is truly necessary since the release process is always managed internally by OTGS and, IMHO there's no need for anyone else to be burdened by it (but as always, I'm open to being convinced otherwise).

**Comment by baizmandesign** [comment]
@zaantar Yes, I wasn't suggesting public documentation for this, only that internal documentation be created (if necessary) for the sake of completeness or in case you ever win the lottery and leave OTGS!

**Comment by zaantar** [comment]
@baizmandesign not sure that would be enough of a motivation to leave OTGS ;) In any case, no worries, we got it covered!

**Pull request merged**